extensions/amp-skimlinks/0.1/link-rewriter/link-rewriter-manager.js

```diff
@@ -0,0 +1,280 @@
+/**
+ * Copyright 2018 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {AmpEvents} from '../../../../src/amp-events';
+import {EVENTS, PRIORITY_META_TAG_NAME} from './constants';
+import {LinkRewriter} from './link-rewriter';
+import {Priority} from '../../../../src/service/navigation';
+import {Services} from '../../../../src/services';
+
+
+/**
+ * LinkRewriterManager works together with LinkRewriter to allow rewriting
+ * links at click time. E.g: Replacing a link by its affiliate version only if
+ * the link can be monetised.
+ *
+ * A page can have multiple LinkRewriter running at the same time but only one
+ * LinkRewriterManager instance.
+ *
+ * LinkRewriterManager class is in charge of:
+ * - Keeping track of all the registered linkRewriters.
+ *
+ * - Notifying all the link rewriters when the DOM has changed (potential new
+ *   links on the page that LinkRewriters need to resolve ahead of the click).
+ *
+ * - Managing which LinkRewriter has the priority to replace a link
+ *   (based on configurable priority list).
+ *
+ * - Watching for anchor clicks and allowing the top priority LinkRewriter to
+ *   execute a potential "href" replacement. At the moment, only one
+ *   LinkRewriter can mutate the link. If the top priority LinkRewriter decides
+ *   to not replace the link, the second top priority LinkRewriter will be
+ *   given a chance. The iteration continues until one LinkRewriter mutates the
+ *   link or we reach the end of the list.
+ *
+ * - Sending a click event to listeners to be able to track the click
+ *   even if the url has not been replaced. (No anchor mutation is allowed
+ *   in the listener handler).
+ */
+export class LinkRewriterManager {
```

Comment:
how is this class different from the one in navigation.js

Reply:
The two solutions are slightly alike since LinkRewriterManager was created when the anchor mutation system didn't exist yet in navigation.js.
However, LinkRewriterManager & LinkRewriter are a bit more advanced/specific than navigation.js. It contains features that would be useful to services similar to Skimlinks:
- Watching for new links inserted in the page, allowing to compute the replacement URL asynchronously.
- Allowing only one LinkRewriter to change the URL based on a priority list configurable by the publisher when running multiple link rewriter in the page (multiple affiliate services for example) 
- Propagate a click event for tracking non-affiliated clicks.
- Keep tracks of which links are allowed to be rewritten based on publisher configuration. (`linkSelector` option)

Those are basic requirements for skimlinks that we have identified has potentially common to other similar services.

Reply:
I see the difference now. I'll let @alabiaga to decide if he think anything valuable to be moved to navigation.js